### PR TITLE
Project: Members-only (Ruby on Rails)

### PR DESCRIPTION
**Pull request by M-Alamgir**
In this project, I built an Blog Post App where your members can write anonymous posts. If logged-in, members can see who the author of a post is but, outside, they can only see the story and wonder who wrote it.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,24 +1,92 @@
-# README
+# Project: Members Only
 
-This README would normally document whatever steps are necessary to get the
-application up and running.
+In this project, I build an Blog Post App where your members can write anonymous posts. If logged-in, members can see who the author of a post is but, outside, they can only see the story and wonder who wrote it.
```
Comment (enaburekhan) [inline README.md]: Replace `build` with `built`. And `an` with `a` in line 3.

File: app/views/layouts/application.html.erb [hunk 2]
```diff
@@ -7,10 +7,28 @@
...
+    <header>
+    <% if current_user %>
+      Hi <%= current_user.name %>
+      |
+      <%= link_to 'Log out', logout_path, method: :delete %>
+      <% else %>
+      
+      <%= link_to 'log in', login_path %>
+
+      <% end %>
```
Comment (enaburekhan) [inline app/views/layouts/application.html.erb]: Move this logic to a helper. Views are only for rendering information that became available to them from the corresponding controllers. And not for generating them.

File: app/views/posts/index.html.erb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  </thead>
+
+  <tbody>
+    <% @posts.each do |post| %>
+      <div class="post-list">
+        <h4 class="post-title"><%= post.title %></h4>
+        <h6><%= post.body %></h6>
+        <% if current_user %>
+        <p class="post-title">posted by <%= post.user.name %></p>
+        <% end %>
```
Comment (enaburekhan) [inline app/views/posts/index.html.erb]: Move this logic to a helper.

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:wrench::hammer::full_moon:
**Hi Alamgir!
Great job so far:clap::clap:.**

**Major highlights**
- The PR is very descriptive

However, we still need to make a few changes to get the project approved:muscle:.

**To improve**

- [x] Linters

- Add .rubocop.yml and .stylelintrc.json config file to the project root directory. They are currently missing. Use this [link](https://github.com/microverseinc/linters-config/tree/master/ror) to get the configs.

- [x] App behavior

- User can sign up ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_up_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/)). I had to follow the instructions in the readme to create a new user from the console to be able to login. I should be able to sign up from the UI and then login. Kindly fix this.

- [x] Authentication and User setup

- The application uses the `devise` gem for authentication ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/bcrypt_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)). I should be able to see the devise authentication in the model. Kindly fix this.

- [x] PostsController

- it uses `before_action` to restrict access to the `:new` and `:create` methods (uses ApplicationController helper method) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)). So inside the helper, we can implement something similar to this:
```ruby
def signed_in_only!
  # current_user in ApplicationController must be defined as helper method to make it work
  redirect_to  new_user_session_path unless current_user

  ## alternative solution:
  # if signed_in? is defined in ApplicationController 
  redirect_to new_user_session_path unless signed_in?
end
```
Then we can add this `before_action :signed_in_only!, only: [:new, :create]` to the top of the `PostsController`

- [x] View

- It has sign up and sign in forms created by the devise gem ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_up_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49722686/)). Devise is currently missing in view. Kindly also fix this.
- Move all the highlighted logic in the PR below to a helper. There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) if you need help on how to use a helper in rails. See [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright

- Remove commented code from `index.html.erb` under post. It makes the codebase to be untidy and difficult to read and understand. Use git history if you must view previous codes.

- [x] Make all the changes highlighted in the PR below.

Make the above-required changes and when you are ready, request another review.

Do have questions/comments? kindly drop them :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy hacking:blush:
Eric Enaburekhan.

File: README.md (hunk 1 above)
Comment (M-Alamgir) [reply README.md]: Done

File: app/views/layouts/application.html.erb (hunk 2 above)
Comment (M-Alamgir) [reply app/views/layouts/application.html.erb]: Done

File: app/views/posts/index.html.erb (hunk 3 above)
Comment (M-Alamgir) [reply app/views/posts/index.html.erb]: Done

**Review by elisha2kyakpo1 — APPROVED** [review]
# Status: Approved🎉✔💪
Job well done for implementing all the requirements requested by the previous reviewer!
The UI looks good, though you can make it look even nicer.
You can merge the pr.
Cheers.

**Pull request merged**